### PR TITLE
Adds the restricted damerau-levenshtein distance

### PR DESCRIPTION
**Pull request by NicklasXYZ**
Adds a function to compute the Restricted Damerau-Levenshtein distance also called the Optimal String Alignment (OSA) distance.

### CODE REVIEW

File: src/edit_distance/osa.gleam [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+fn string_to_dict(str: String) -> dict.Dict(Int, String) {
+  // Convert the string to a list of strings and build a dictionary
+  let string_graphemes = string.to_graphemes(str)
+  let string_dict =
+    list.index_fold(string_graphemes, [], fn(acc, item, i) {
+      [#(i, item), ..acc]
+    })
+
+  // Create a dictionary where keys are indices and values are strings
+  dict.from_list(string_dict)
```
Comment (giacomocavalieri) [inline src/edit_distance/osa.gleam]: This creates an intermediate lists in memory and has to go through it to build the final dictionary. Could you build the dictionary as you go using `dict.new()` as the accumulator and inserting the values as you go along?

File: src/edit_distance/osa.gleam [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  // Compute the final distances between substrings
+  let distance_dict =
+    compute_distances(one_dict, other_dict, distance_dict, n, m)
+
+  // Return the final distance between the full strings
+  let assert Ok(result) = dict.get(distance_dict, #(n, m))
+  result
+}
+
+fn string_to_dict(str: String) -> dict.Dict(Int, String) {
```
Comment (giacomocavalieri) [inline src/edit_distance/osa.gleam]: I'd prefer the `Dict` type to be unqualified:
```suggestion
fn string_to_dict(str: String) -> Dict(Int, String) {
```

File: src/edit_distance/osa.gleam [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            [#(#(i, j), i), ..inner_acc]
+          }
+          // Initialize the rest of the matrix with 0.0
+          False, False -> [#(#(i, j), 0), ..inner_acc]
+        }
+      })
+    })
+
+  // Convert the initialized matrix to a dictionary
+  dict.from_list(matrix)
```
Comment (giacomocavalieri) [inline src/edit_distance/osa.gleam]: Same as before, I'd build the dictionary as I go, use `use` and shadow the `acc` variable:
```gleam
use acc, i <- list.fold(list.range(0, n), dict.new())
use acc, j <- list.fold(list.range(0, m), acc)
todo as "insert the values in the dict"
```

File: src/edit_distance/osa.gleam [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+      // Check for transpositions to update the distances
+      let final_d =
+        dict.insert(
+          updated_d,
+          #(i, j),
+          check_transposition(one, other, updated_d, i, j),
+        )
+      final_d
+    })
+  })
```
Comment (giacomocavalieri) [inline src/edit_distance/osa.gleam]: I'd use `use` and shadow the accumulator here as well:
```gleam
use acc, i <- list.fold(list.range(1, n), d)
use acc, j <- list.fold(list.range(1, m), acc)
todo as "rest..."
```

**Review by giacomocavalieri — COMMENTED** [review]
I find a bit hard to follow the algorithm, could you add more comments in the code explaining what those dictionaries you're using represent and why they're used?

**Comment by NicklasXYZ** [comment]
> I find a bit hard to follow the algorithm, could you add more comments in the code explaining what those dictionaries you're using represent and why they're used?

The two first dictionaries used index the input strings as index-letter pairs.

The main dictionary I am traversing/using essentially represents the table/matrix constructed using the recursive formula given on Wikipedia.

I'll add more comments, when I have some additional time to spare :)

**Comment by NicklasXYZ** [comment]
I tried to address all your comments. Hopefully I made the implementation more clear :)

File: src/edit_distance/osa.gleam [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+  let string_graphemes = string.to_graphemes(str)
+
+  // Build a list of tuples where each tuple contains an index and the corresponding character
+  let string_dict =
+    list.index_fold(string_graphemes, [], fn(acc, item, i) {
+      [#(i, item), ..acc]
+    })
+
+  // Convert the list of index-character tuples into a dictionary
+  dict.from_list(string_dict)
```
Comment (giacomocavalieri) [inline src/edit_distance/osa.gleam]: Building an intermediate list is wasteful, can you build the dictionary directly instead of building an intermediate list?
```gleam
// pseudo code
use acc, item, i <- list.index_fold(..., from: dict.new())
dict.insert(acc, i, item)
```

File: src/edit_distance/osa.gleam [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+// The function retrieves the cost using the tuple (row_index, column_index) as the key. 
+// The value corresponds to the distance calculated between the substrings: 
+// - 'one' (from position 0 to row_index)
+// - 'other' (from position 0 to column_index) 
+// Since the matrix (and thus the dictionary) is processed iteratively (row-wise and column-wise), 
+// the value at index (row_index, column_index), where row_index < i and column_index < j, 
+// will already be present in the matrix by the time we calculate the value for entry (i, j). We 
+// can thus access this entry with confidence
+fn get_cost(d: Dict(#(Int, Int), Int), row: Int, column: Int) -> Int {
+  let assert Ok(cost) = dict.get(d, #(row, column))
```
Comment (giacomocavalieri) [inline src/edit_distance/osa.gleam]: Let's not assert here and instead assert where it's called, that way it can be easier to see what it's called with and why it cannot fail

File: src/edit_distance/osa.gleam [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    use acc_d, j <- list.fold(list.range(0, n), d)
+    dict.insert(acc_d, #(0, j), j)
+  }
+
+  // Iterate over the matrix to calculate the minimum costs for all substrings
+  use acc_d, i <- list.fold(list.range(1, m), d)
+  use inner_d, j <- list.fold(list.range(1, n), acc_d)
+  // Calculate the cost for this position (i, j)
+  let cost = compute_cost(one, other, inner_d, i, j)
+  // Update the distance dictionary with the calculated cost
```
Comment (giacomocavalieri) [inline src/edit_distance/osa.gleam]: Since you're not using `acc_d` in the inner loop could you rename `inner_d` to `acc_d` to shadow it?

File: src/edit_distance/osa.gleam [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    dict.insert(acc_d, #(i, 0), i)
+  }
+  let d = {
+    use acc_d, j <- list.fold(list.range(0, n), d)
+    dict.insert(acc_d, #(0, j), j)
+  }
+
+  // Iterate over the matrix to calculate the minimum costs for all substrings
+  use acc_d, i <- list.fold(list.range(1, m), d)
+  use inner_d, j <- list.fold(list.range(1, n), acc_d)
```
Comment (giacomocavalieri) [inline src/edit_distance/osa.gleam]: I'd like this to be as efficient as possible, calling `list.range` creates an entire list in memory to fold over it. Here you could define your own `fold_range` function that looks like this:
```gleam
fn fold_range(from start: Int, to end: Int, acc acc: a, do fun: fn(a, Int) -> a) -> a {
  // This calls recursively from start to end without
  // having to build a possibly big list in memory
}
```
And then this code would become:
```gleam
use acc, i <- fold_range(from: 1, to: m, acc: d)
use acc, j <- fold_range(from: 1, to: n, acc: acc)
...
```

**Review by giacomocavalieri — COMMENTED** [review]
Thanks for adding more comments! I've left a couple more notes inline

**Comment by giacomocavalieri** [comment]
Lovely, thank you!

**Comment by NicklasXYZ** [comment]
No problemo and thanks for the comments :)

In case you are thinking about creating a new release, I hope to try implementing some of the other edit distance functions before the end of the week, in case you want to wait.
